src/storage.rs

```diff
@@ -0,0 +1,261 @@
+use fastly::kv_store::{InsertMode, KVStoreError, LookupResponse};
+use fastly::KVStore;
+use std::time::Duration;
+
+// the amount of time to wait before deleting an item after its expiration
+// is reached. keeping expired items around allows their sequencing
+// information to be reused if they are later updated prior to deletion.
+// this helps reduce the chance of sequences restarting, which can be
+// disruptive to message delivery.
+const LINGER: Duration = Duration::from_secs(60 * 60 * 24);
+
+const WRITE_TRIES_MAX: usize = 5;
+
+#[derive(Debug)]
+pub enum StorageError {
+    StoreNotFound,
+    TooManyRequests,
+    InvalidMetadata,
+    KVStore(KVStoreError),
+}
+
+pub struct RetainedVersion {
+    pub generation: u64,
+    pub seq: u64,
+}
+
+pub struct RetainedMessage {
+    pub ttl: Option<Duration>,
+    pub data: Vec<u8>,
+}
+
+pub struct RetainedSlot {
+    pub version: RetainedVersion,
+    pub message: Option<RetainedMessage>,
+}
+
+#[derive(Debug, Default, serde::Deserialize, serde::Serialize)]
+struct Metadata {
```

Comment:
Would it be helpful to version this for the future?

Reply:
I tend to prefer evolving formats without version numbers, which we should have room to do here since it's JSON. Can always slap on a version field later if that doesn't work out.